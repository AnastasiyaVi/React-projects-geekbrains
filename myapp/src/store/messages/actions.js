import { AUTHORS } from "../../components/Author";
import { database } from "../../services/firebase";

export const SET_MESSAGES = "MESSAGES::SET_MESSAGES";

export const addMessage = (chatId, text, author) => async () => {
  setTimeout(() => {
    database.ref("messages").push({
      chatId,
      text: "Я-робот",
      author: AUTHORS.BOT,
    });
  }, 1000);

  database.ref("messages").push({
    chatId,
    text,
    author,
  });
};

export const deleteMessage = (chatId, id) => () => {
  database.ref("messages").child(id).remove();
};

export const getMessagesFromFirebase = () => async (dispatch) => {
  let messages = {};

  // подписываемся на получение, изменение сообщений, только в этом месте диспатчим(меняем) стейт redux
  // так как в realtime database подключение на сокетах, нам не нужно в функциях addMessage и addChat менять стейт
  database.ref("messages").on("value", (snapshot) => {
    snapshot.forEach((item) => {
      const message = item.val();
      if (!Array.isArray(messages?.[message.chatId])) {
        messages[message.chatId] = [];
      }
      messages[message.chatId].push({
        ...message,
        id: item.key,
      });
    });

    dispatch({
      type: SET_MESSAGES,
      payload: messages,
    });

    messages = {};
  });
};
