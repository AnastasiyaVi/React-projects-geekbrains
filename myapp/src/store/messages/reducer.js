import { SET_MESSAGES } from "./actions";

// statuses - 'idle' | 'fetching' | 'succeeded' | 'failed',
// error - string | null

const initialState = {
  messages: {},
  status: "idle",
  error: null,
};

export const messagesReducer = (state = initialState, { type, payload }) => {
  switch (type) {
    case SET_MESSAGES: {
      return {
        ...state,
        messages: payload,
      };
    }
    default:
      return state;
  }
};
