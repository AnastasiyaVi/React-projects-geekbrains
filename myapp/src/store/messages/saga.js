import { put, takeEvery, call } from 'redux-saga/effects'
import { AUTHORS } from "../../components/Author";
import { ADD_MESSAGE, addMessage } from './actions'

const delay = time => new Promise(resolve => setTimeout(resolve, time)); //Базовая задержка со значением

// с помощью yield указываем, какие действия необходимо совершить
function* addBotMessage({ payload }) {
    if (payload.author === AUTHORS.HUMAN) {
        yield call(delay, 1000); //call - вызывает переданную функцию, 
        yield put(addMessage(payload.chatId, 'Я-робот', AUTHORS.BOT)) //put - диспатчит в стор переданный экшен
    }
}

export default function* watchAddMessage() {
    yield takeEvery(ADD_MESSAGE, addBotMessage); //указываем милдвару какие экшены необходимо перехватить, При диспатче этого экшена   будет вызвана функция-генератор addBotMessage
}