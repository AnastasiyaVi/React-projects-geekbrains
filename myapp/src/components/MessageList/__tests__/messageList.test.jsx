import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { SET_MESSAGES } from "../../../store/messages/actions";
import { MessageList } from "../index";
const { render } = require("@testing-library/react");
const { Provider } = require("react-redux");

const middlewares = [thunk];
// Используем библиотеку для создания мокового стора
const mockStore = configureStore(middlewares);

// описание
it("Message: matches snapshot", async () => {
  const initialState = {
    chats: [{ id: "chat-1", name: "test chat" }],
    messages: {
      messages: [],
    },
  };

  const store = mockStore(initialState);

  const addMessage = () => ({
    type: SET_MESSAGES,
    payload: [
      {
        id: "dasndj1ddsjk",
        chatId: "chat-1",
        text: "test text",
        author: "test",
      },
    ],
  });
  store.dispatch(addMessage());

  // Рендерим компонент сообщений, обернув в Provider, который симулирует redux
  const component = render(
    <Provider store={store}>
      <MessageList chatId="chat-1" chatExists />
    </Provider>
  );

  // Меняем значения в пропсе и сравниваем снапшоты на валидность
  expect(component).toMatchSnapshot();
});
