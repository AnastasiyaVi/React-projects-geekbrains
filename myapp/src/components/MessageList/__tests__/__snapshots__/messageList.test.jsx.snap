// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message: matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="m-0 p-0 card"
      >
        <div
          class="ContentProfile d-flex flex-row justify-content-start align-items-center card-header"
        >
          <p
            class="ContactProfileTxt m-0 p-0"
          />
        </div>
        <div
          class="Messages card-body"
        >
          <div
            class="Scrollable"
          />
        </div>
        <div
          class="card-footer"
        >
          <form>
            <div
              class="messages_form"
            >
              <textarea
                aria-label="maximum height"
                class="message_text"
                placeholder="Сообщение"
                rows="1"
                style="width: 200px; height: 0px; overflow: hidden;"
              />
              <textarea
                aria-hidden="true"
                class="message_text"
                readonly=""
                style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 200px; padding: 0px;"
                tabindex="-1"
              />
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium message_btn css-sghohy-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="submit"
                value=""
              >
                Отправить
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="m-0 p-0 card"
    >
      <div
        class="ContentProfile d-flex flex-row justify-content-start align-items-center card-header"
      >
        <p
          class="ContactProfileTxt m-0 p-0"
        />
      </div>
      <div
        class="Messages card-body"
      >
        <div
          class="Scrollable"
        />
      </div>
      <div
        class="card-footer"
      >
        <form>
          <div
            class="messages_form"
          >
            <textarea
              aria-label="maximum height"
              class="message_text"
              placeholder="Сообщение"
              rows="1"
              style="width: 200px; height: 0px; overflow: hidden;"
            />
            <textarea
              aria-hidden="true"
              class="message_text"
              readonly=""
              style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 200px; padding: 0px;"
              tabindex="-1"
            />
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium message_btn css-sghohy-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="submit"
              value=""
            >
              Отправить
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
